File: scripts/chronicle-keeper/capsule-scroll-generator.js
#!/usr/bin/env node
/**
 * Capsule Scroll Generator
 * 
 * Generates Capsule Scrolls for preserving conversation context when hitting archive walls.
 * Integrates with Chronicle Keeper system for automated generation.
 */

const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

class CapsuleScrollGenerator {
    constructor() {
        this.projectRoot = process.env.GITHUB_WORKSPACE || process.cwd();
        this.capsulesDir = path.join(this.projectRoot, 'capsules');
        this.templatesDir = path.join(this.capsulesDir, 'templates');
        this.activeDir = path.join(this.capsulesDir, 'active');
        this.archivedDir = path.join(this.capsulesDir, 'archived');
        
        this.stats = {
            generated: 0,
            archived: 0,
            errors: 0
        };
    }

    /**
     * Generate a new Capsule Scroll from parsed GitHub content
     * @param {Object} parsedContent - Parsed content from Chronicle Keeper
     * @returns {Object} Generation result with filename and content
     */
    async generateCapsuleScroll(parsedContent) {
        try {
            const scrollType = this.determineScrollType(parsedContent);
            const template = await this.loadTemplate(scrollType);
            const populated = this.populateTemplate(template, parsedContent);
            const filename = this.generateFilename(parsedContent);
            
            await this.writeScrollFile(filename, populated);
            
            this.stats.generated++;
            return {
                success: true,
                filename,
                content: populated,
                scrollType,
                stats: this.stats
            };
            
        } catch (error) {
            this.stats.errors++;
            return {
                success: false,
                error: error.message,
                stats: this.stats
            };
        }
    }

    /**
     * Determine what type of scroll to generate based on content
     * @param {Object} parsedContent - Parsed GitHub content
     * @returns {string} Template type to use
     */
    determineScrollType(parsedContent) {
        // Look for specific decision-making context (not just words)
        const decisionKeywords = ['choose between', 'option a', 'option b', 'alternatives', 'pros and cons'];
        const contentText = (parsedContent.body || '').toLowerCase();
        
        const hasStrongDecisionContext = decisionKeywords.some(keyword => 
            contentText.includes(keyword)
        );
        
        if (hasStrongDecisionContext) {
            return 'decision-capsule';
        }
        
        // For archive wall and continuity contexts, use arc template
        if (contentText.includes('archive wall') || contentText.includes('continuity') || 
            contentText.includes('conversation') || parsedContent.title.includes('🧠📜')) {
            return 'arc-capsule';
        }
        
        // Default to arc capsule for conversation preservation
        return 'arc-capsule';
    }

    /**
     * Load template file
     * @param {string} templateType - Type of template (arc-capsule or decision-capsule)
     * @returns {string} Template content
     */
    async loadTemplate(templateType) {
        const templatePath = path.join(this.templatesDir, `${templateType}.md`);
        
        if (!fs.existsSync(templatePath)) {
            throw new Error(`Template not found: ${templatePath}`);
        }
        
        return fs.readFileSync(templatePath, 'utf8');
    }

    /**
     * Populate template with actual content
     * @param {string} template - Template string with placeholders
     * @param {Object} parsedContent - Data to populate template with
     * @returns {string} Populated template
     */
    populateTemplate(template, parsedContent) {
        const now = new Date().toISOString();
        const dateStr = now.split('T')[0]; // YYYY-MM-DD format
        
        // Create replacement map
        const replacements = {
            '{ARC_NAME}': parsedContent.title || 'Untitled Conversation Arc',
            '{DECISION_NAME}': parsedContent.title || 'Untitled Decision',
            '{START_DATE}': parsedContent.created_at ? parsedContent.created_at.split('T')[0] : dateStr,
            '{END_DATE}': dateStr,
            '{DECISION_DATE}': parsedContent.created_at ? parsedContent.created_at.split('T')[0] : dateStr,
            '{CONVERSATION_MODE}': this.inferConversationMode(parsedContent),
            '{ISSUE_OR_PR_CONTEXT}': parsedContent.url || 'Context not available',
            '{PARTICIPANTS}': parsedContent.author || 'Unknown',
            '{CREATION_DATE}': now,
            '{UPDATE_DATE}': now,
            '{STATUS}': 'Active',
            '{ACTIVE_OR_ARCHIVED}': 'Active'
        };

        // Extract features, decisions, artifacts from content
        this.extractContentElements(parsedContent, replacements);
        
        // Replace all placeholders, including ones that weren't explicitly set
        let populated = template;
        
        // First pass: replace explicit values
        for (const [placeholder, value] of Object.entries(replacements)) {
            populated = populated.replace(new RegExp(placeholder.replace(/[{}]/g, '\\$&'), 'g'), value);
        }
        
        // Second pass: replace any remaining placeholders with sensible defaults
        const defaultReplacements = {
            '{DESCRIPTION_OF_THE_PROBLEM_OR_CHOICE_THAT_NEEDED_TO_BE_MADE}': 'Archive wall context preservation challenge - need to maintain conversation continuity when hitting context limits.',
            '{OPTION_A_NAME}': 'Manual Context Summarization',
            '{OPTION_B_NAME}': 'Capsule Scrolls System',
            '{OPTION_C_NAME}': 'Enhanced TLDL Integration',
            '{PROS_LIST}': 'Preserves context, enables continuation, saves developer time',
            '{CONS_LIST}': 'Requires initial setup, needs maintenance',
            '{IMPACT_ASSESSMENT}': 'High positive impact on developer experience',
            '{CHOSEN_OPTION}': 'Capsule Scrolls System',
            '{EXPLANATION_OF_WHY_THIS_OPTION_WAS_CHOSEN}': 'Provides automated, structured approach to context preservation that integrates with existing Chronicle Keeper system.',
            '{FACTOR_1}': 'Addresses immediate archive wall problem',
            '{FACTOR_2}': 'Builds on existing infrastructure',
            '{FACTOR_3}': 'Aligns with "Save the Butts" philosophy',
            '{IMPLEMENTATION_TIMELINE}': 'Current implementation phase',
            '{CHANGES_REQUIRED}': 'Add capsules directory, templates, generator script, Chronicle Keeper integration',
            '{HOW_TO_MEASURE_SUCCESS}': 'Successful context preservation across conversation boundaries',
            '{WHAT_TO_DO_IF_IT_GOES_WRONG}': 'Fall back to manual TLDL creation and context summarization',
            '{ACTION_ITEM_1}': 'Complete Capsule Scroll generator implementation',
            '{ACTION_ITEM_2}': 'Integrate with Chronicle Keeper workflow',
            '{ACTION_ITEM_3}': 'Test with real conversation scenarios',
            '{PARTICIPANT_1}': `${parsedContent.author || 'contributor'} — Primary implementer`,
            '{PARTICIPANT_2}': 'Chronicle Keeper — Automation system',
            '{RUNNING_JOKE_OR_MEME}': 'Save the Butts continuity ritual',
            '{PROJECT_SPECIFIC_TERMINOLOGY}': 'Archive wall, molting ritual for code, Chronicle Keeper',
            '{NEXT_REVIEW_DATE}': new Date(Date.now() + 7 * 24 * 60 * 60 * 1000).toISOString().split('T')[0] // 1 week from now
        };
        
        for (const [placeholder, defaultValue] of Object.entries(defaultReplacements)) {
            populated = populated.replace(new RegExp(placeholder.replace(/[{}]/g, '\\$&'), 'g'), defaultValue);
        }
        
        return populated;
    }

    /**
     * Extract content elements from parsed GitHub content
     * @param {Object} parsedContent - Parsed GitHub content
     * @param {Object} replacements - Replacement map to populate
     */
    extractContentElements(parsedContent, replacements) {
        const body = parsedContent.body || '';
        
        // Try to extract key elements from the content
        const lines = body.split('\n');
        
        // Look for features/decisions/artifacts in bullet points or numbered lists
        const features = [];
        const decisions = [];
        const artifacts = [];
        const glyphs = [];
        const threads = [];
        
        for (const line of lines) {
            const trimmed = line.trim();
            
            // Extract bullet points and numbered items
            if (trimmed.match(/^[-*+]\s+/) || trimmed.match(/^\d+\.\s+/)) {
                const content = trimmed.replace(/^[-*+\d.]\s*/, '');
                
                if (content.toLowerCase().includes('feature') || content.toLowerCase().includes('implement')) {
                    features.push(content);
                } else if (content.toLowerCase().includes('decision') || content.toLowerCase().includes('choose')) {
                    decisions.push(content);
                } else if (content.includes('http') || content.includes('github.com')) {
                    artifacts.push(content);
                } else if (content.includes('🧠') || content.includes('📜') || content.includes('🍑')) {
                    glyphs.push(content);
                } else if (content.toLowerCase().includes('todo') || content.toLowerCase().includes('unresolved')) {
                    threads.push(content);
                }
            }
        }
        
        // Populate replacements with extracted content or defaults
        replacements['{FEATURE_1}'] = features[0] || 'Archive Wall Context Preservation';
        replacements['{FEATURE_2}'] = features[1] || 'Capsule Scroll Generation';
        replacements['{FEATURE_3}'] = features[2] || 'Chronicle Keeper Integration';
        
        replacements['{BRIEF_DESCRIPTION}'] = 'Implementation to preserve conversation context when hitting archive walls';
        
        replacements['{DECISION_1}'] = decisions[0] || 'Implement Capsule Scroll system for context preservation';
        replacements['{DECISION_2}'] = decisions[1] || 'Use existing daily ledger format as template base';
        replacements['{DECISION_3}'] = decisions[2] || 'Integrate with Chronicle Keeper for automation';
        
        replacements['{ARTIFACT_1}'] = artifacts[0] || `[Issue #${parsedContent.number || 'unknown'}](${parsedContent.url || '#'})`;
        replacements['{ARTIFACT_2}'] = artifacts[1] || '[Capsule Scrolls README](../README.md)';
        replacements['{ARTIFACT_3}'] = artifacts[2] || '[Chronicle Keeper Integration](../scripts/chronicle-keeper/)';
        
        replacements['{GLYPH_1}'] = glyphs[0] || '🧠📜 — Archive Wall Continuity Ritual';
        replacements['{GLYPH_2}'] = glyphs[1] || '🍑 — Save the Butts clause'; 
        replacements['{GLYPH_3}'] = glyphs[2] || '🧙‍♂️ — Chronicle Keeper automation';
        
        replacements['{CONTEXT}'] = 'Preserves development wisdom and conversation continuity';
        
        replacements['{THREAD_1}'] = threads[0] || 'Test automated generation workflow';
        replacements['{FEATURE_1}'] = features[0] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{FEATURE_1}'];
        replacements['{FEATURE_2}'] = features[1] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{FEATURE_2}'];
        replacements['{FEATURE_3}'] = features[2] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{FEATURE_3}'];
        
        replacements['{BRIEF_DESCRIPTION}'] = CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{BRIEF_DESCRIPTION}'];
        
        replacements['{DECISION_1}'] = decisions[0] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{DECISION_1}'];
        replacements['{DECISION_2}'] = decisions[1] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{DECISION_2}'];
        replacements['{DECISION_3}'] = decisions[2] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{DECISION_3}'];
        
        replacements['{ARTIFACT_1}'] = artifacts[0] || `[Issue #${parsedContent.number || 'unknown'}](${parsedContent.url || '#'})`;
        replacements['{ARTIFACT_2}'] = artifacts[1] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{ARTIFACT_2}'];
        replacements['{ARTIFACT_3}'] = artifacts[2] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{ARTIFACT_3}'];
        
        replacements['{GLYPH_1}'] = glyphs[0] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{GLYPH_1}'];
        replacements['{GLYPH_2}'] = glyphs[1] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{GLYPH_2}']; 
        replacements['{GLYPH_3}'] = glyphs[2] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{GLYPH_3}'];
        
        replacements['{CONTEXT}'] = CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{CONTEXT}'];
        
        replacements['{THREAD_1}'] = threads[0] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{THREAD_1}'];
        replacements['{THREAD_2}'] = threads[1] || CapsuleScrollGenerator.DEFAULT_REPLACEMENTS['{THREAD_2}'];
        replacements['{THREAD_3}'] = threads[2] || 'Create proof-of-concept scroll for this implementation';
        
        // Generate re-entry spell
        const reentrySpell = this.generateReentrySpell(parsedContent);
        replacements['{THREE_SENTENCE_SUMMARY_OF_THE_CONVERSATION_ARC_THAT_ENABLES_QUICK_CONTEXT_RESTORATION}'] = reentrySpell;
        
        // Add reason/status defaults
        replacements['{REASON}'] = 'implementation in progress';
        replacements['{STATUS}'] = 'pass';
        
        // Add links
        replacements['{LINK_1}'] = parsedContent.url || '#';
        replacements['{LINK_2}'] = '../README.md';
        replacements['{LINK_3}'] = '../scripts/chronicle-keeper/';
        
        replacements['{TLDL_ENTRY_1}'] = '[Chronicle Keeper Implementation](../../TLDL/entries/TLDL-2025-08-07-ChronicleKeeperImplementation.md)';
        replacements['{TLDL_ENTRY_2}'] = '[Daily Ledger Format](../../docs/daily-ledger/2025-08-18.md)';
        replacements['{TLDL_LINK_1}'] = '../../TLDL/entries/TLDL-2025-08-07-ChronicleKeeperImplementation.md';
        replacements['{TLDL_LINK_2}'] = '../../docs/daily-ledger/2025-08-18.md';
        
        replacements['{ISSUE_1}'] = `Issue #${parsedContent.number || 'unknown'}`;
        replacements['{PR_1}'] = 'PR for Capsule Scrolls Implementation';
        replacements['{ISSUE_LINK_1}'] = parsedContent.url || '#';
        replacements['{PR_LINK_1}'] = '#';
    }

    /**
     * Generate a re-entry spell (3-sentence summary)
     * @param {Object} parsedContent - Parsed GitHub content  
     * @returns {string} Three-sentence summary
     */
    generateReentrySpell(parsedContent) {
        const title = parsedContent.title || 'Archive Wall Conversation';
        const author = parsedContent.author || 'the team';
        const context = parsedContent.body ? 
            parsedContent.body.substring(0, 100).replace(/\n/g, ' ') + '...' : 
            'context preservation implementation';
        
        return `${author} initiated work on "${title}" to address archive wall limitations in long conversations. ` +
               `The implementation focuses on ${context} using the Chronicle Keeper system. ` +
               `This Capsule Scroll preserves the conversation context to enable smooth re-entry when context limits are reached.`;
    }

    /**
     * Infer conversation mode from content
     * @param {Object} parsedContent - Parsed GitHub content
     * @returns {string} Conversation mode
     */
    inferConversationMode(parsedContent) {
        const body = (parsedContent.body || '').toLowerCase();
        
        if (body.includes('keeper') || body.includes('chronicle')) {
            return 'Keeper-plus mode';
        } else if (body.includes('debug') || body.includes('troubleshoot')) {
            return 'Debug & investigation';
        } else if (body.includes('feature') || body.includes('implement')) {
            return 'Feature development';
        } else {
            return 'General development discussion';
        }
    }

    /**
     * Generate filename for the scroll
     * @param {Object} parsedContent - Parsed GitHub content
     * @returns {string} Filename
     */
    generateFilename(parsedContent) {
        const now = new Date();
        const dateStr = now.toISOString().split('T')[0]; // YYYY-MM-DD
        
        let title = parsedContent.title || 'untitled-scroll';
        title = title.replace(/[🧠📜🍑🧙‍♂️]/g, ''); // Remove emojis
        title = title.replace(/[^a-zA-Z0-9\s-]/g, ''); // Remove special chars
        title = title.replace(/\s+/g, '-'); // Replace spaces with hyphens
        title = title.toLowerCase();
        title = title.substring(0, 50); // Limit length
        
        return `${dateStr}-${title}.md`;
    }

    /**
     * Write scroll content to file
     * @param {string} filename - Name of file to create
     * @param {string} content - Content to write
     */
    async writeScrollFile(filename, content) {
        const filepath = path.join(this.activeDir, filename);
        
        // Ensure directory exists
        if (!fs.existsSync(this.activeDir)) {
            fs.mkdirSync(this.activeDir, { recursive: true });
        }
        
        fs.writeFileSync(filepath, content, 'utf8');
    }

    /**
     * Archive an active scroll (move from active/ to archived/)
     * @param {string} filename - Name of file to archive
     * @returns {Object} Result of archiving operation
     */
    async archiveScroll(filename) {
        try {
            const activePath = path.join(this.activeDir, filename);
            const archivedPath = path.join(this.archivedDir, filename);
            
            if (!fs.existsSync(activePath)) {
                throw new Error(`Active scroll not found: ${filename}`);
            }
            
            // Ensure archived directory exists
            if (!fs.existsSync(this.archivedDir)) {
                fs.mkdirSync(this.archivedDir, { recursive: true });
            }
            
            // Move file
            fs.renameSync(activePath, archivedPath);
            this.stats.archived++;
            
            return {
                success: true,
                message: `Archived scroll: ${filename}`,
                stats: this.stats
            };
            
        } catch (error) {
            this.stats.errors++;
            return {
                success: false,
                error: error.message,
                stats: this.stats
            };
        }
    }

    /**
     * List all scrolls (active and archived)
     * @returns {Object} Lists of scrolls by status
     */
    listScrolls() {
        const activeScrolls = fs.existsSync(this.activeDir) ? 
            fs.readdirSync(this.activeDir).filter(f => f.endsWith('.md')) : [];
        
        const archivedScrolls = fs.existsSync(this.archivedDir) ? 
            fs.readdirSync(this.archivedDir).filter(f => f.endsWith('.md')) : [];
        
        return {
            active: activeScrolls,
            archived: archivedScrolls,
            total: activeScrolls.length + archivedScrolls.length
        };
    }

    /**
     * Get current generation statistics
     * @returns {Object} Current stats
     */
    getStats() {
        return { ...this.stats };
    }
}

// CLI interface for direct usage
if (require.main === module) {
    const generator = new CapsuleScrollGenerator();
    
    // Example usage with mock GitHub issue data
    const mockIssue = {
        number: 57,
        title: "🧠📜 Archive Wall Capsule Scrolls — \"Save the Butts\" Continuity Ritual",
        body: `Ha — double‑joke achievement unlocked. 🏆 The "make this an issue" line *is* already a meta‑gag when you're literally creating… an issue...

**Describe the solution you'd like**  
Introduce **Capsule Scrolls** — a compact, ritualized markdown artifact capturing:  
1. Arc Name  
2. Timeframe  
3. Core Decisions  
4. Key Artifacts & Commits  
5. Glyphs & Running Jokes  
6. Unresolved Threads  
7. *Re‑entry Spell* (3‑sentence context snapshot for scene‑reset)  

Stored alongside the relevant commits or in \`/capsules/\`, linked into the TLDL.`,
        user: { login: "jmeyer1980" },
        created_at: new Date().toISOString(),
        html_url: "https://github.com/jmeyer1980/living-dev-agent/issues/57"
    };
    
    // Parse the mock issue (simulate Chronicle Keeper parsing)
    const parsedContent = {
        type: 'issue',
        number: mockIssue.number,
        title: mockIssue.title,
        body: mockIssue.body,
        author: mockIssue.user.login,
        created_at: mockIssue.created_at,
        url: mockIssue.html_url,
        lore_worthy: true,
        category: 'capsule_scroll'
    };
    
    console.log("📜 Capsule Scroll Generator - Chronicle Keeper Integration");
    console.log("=========================================================");
    
    generator.generateCapsuleScroll(parsedContent).then(result => {
        console.log("🧠📜 Generation Result:");
        console.log(JSON.stringify(result, null, 2));
        
        console.log("\n📊 Current Stats:");
        console.log(generator.getStats());
        
        console.log("\n📋 Scroll Inventory:");
        console.log(generator.listScrolls());
    }).catch(error => {
        console.error("❌ Generation failed:", error);
    });
}

module.exports = CapsuleScrollGenerator;

File: packages/com.twg.the-seed/The Living Dev Agent/scripts/capsule_scroll_generator.py
#!/usr/bin/env python3
"""
Capsule Scroll Generator - Archive Wall Continuity System
Creates conversation context anchors for preserving development wisdom.

Usage:
    python3 capsule_scroll_generator.py --arc-name "Feature Development Arc" --start-date "2025-08-18"
    python3 capsule_scroll_generator.py --from-daily-ledger docs/daily-ledger/2025-08-18.md
"""

import argparse
import os
import sys
import yaml
from datetime import datetime, timedelta
import re
import json

class CapsuleScrollGenerator:
    def __init__(self, project_root=None):
        if project_root is None:
            project_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
        self.project_root = project_root
        self.capsules_dir = os.path.join(project_root, 'capsules')
        self.templates_dir = os.path.join(self.capsules_dir, 'templates')
        self.active_dir = os.path.join(self.capsules_dir, 'active')
        self.archived_dir = os.path.join(self.capsules_dir, 'archived')
        
        # Ensure directories exist
        for directory in [self.capsules_dir, self.templates_dir, self.active_dir, self.archived_dir]:
            os.makedirs(directory, exist_ok=True)
    
    def create_capsule_id(self, arc_name, date_str=None):
        """Generate a unique capsule ID."""
        if date_str is None:
            date_str = datetime.now().strftime('%Y-%m-%d')
        slug = re.sub(r'[^a-zA-Z0-9]+', '-', arc_name.lower()).strip('-')[:30]
        return f"CAPS-{date_str}-{slug}"
    
    def load_template(self):
        """Load the capsule scroll template."""
        template_path = os.path.join(self.templates_dir, 'capsule-scroll-template.md')
        if not os.path.exists(template_path):
            raise FileNotFoundError(f"Template not found: {template_path}")
        
        with open(template_path, 'r', encoding='utf-8') as f:
            return f.read()
    
    def extract_from_daily_ledger(self, ledger_path):
        """Extract capsule scroll data from a daily ledger."""
        if not os.path.exists(ledger_path):
            raise FileNotFoundError(f"Daily ledger not found: {ledger_path}")
        
        with open(ledger_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # Extract metadata from the daily ledger
        data = {
            'arc_name': 'Daily Development Arc',
            'start_date': '',
            'end_date': '',
            'status': 'ongoing',
            'core_decisions': [],
            'artifacts': [],
            'glyphs': [],
            'unresolved_threads': [],
            're_entry_spell': ''
        }
        
        # Parse the daily ledger sections
        if '**Arc:** ' in content:
            arc_match = re.search(r'\*\*Arc:\*\* (.+)', content)
            if arc_match:
                data['arc_name'] = arc_match.group(1).strip()
        
        # Extract date from filename or content
        date_match = re.search(r'(\d{4}-\d{2}-\d{2})', ledger_path)
        if date_match:
            data['start_date'] = date_match.group(1)
            data['end_date'] = date_match.group(1)
        
        # Extract core decisions from "Core prompts / decisions today"
        decisions_match = re.search(r'\*\*Core prompts / decisions today:\*\*\s*\n(.*?)\n\n', content, re.DOTALL)
        if decisions_match:
            decisions_text = decisions_match.group(1)
            for line in decisions_text.split('\n'):
                if line.strip().startswith('-'):
                    decision = line.strip()[1:].strip()
                    data['core_decisions'].append({
                        'title': decision[:50] + '...' if len(decision) > 50 else decision,
                        'context': 'From daily ledger',
                        'outcome': decision,
                        'rationale': 'Daily development decision',
                        'impact': 'To be determined'
                    })
        
        # Extract artifacts from "Key Artifacts & Commits"
        artifacts_match = re.search(r'\*\*Key Artifacts & Commits:\*\*\s*\n(.*?)\n\n', content, re.DOTALL)
        if artifacts_match:
            artifacts_text = artifacts_match.group(1)
            for line in artifacts_text.split('\n'):
                if line.strip().startswith('-') and '[' in line and '](' in line:
                    data['artifacts'].append(line.strip()[1:].strip())
        
        # Extract glyphs from "Glyphs / Running Jokes"
        glyphs_match = re.search(r'\*\*Glyphs / Running Jokes:\*\*\s*\n(.*?)\n\n', content, re.DOTALL)
        if glyphs_match:
            glyphs_text = glyphs_match.group(1)
            for line in glyphs_text.split('\n'):
                if line.strip().startswith('-'):
                    glyph = line.strip()[1:].strip()
                    data['glyphs'].append({
                        'term': glyph.replace('"', ''),
                        'definition': 'Cultural reference from daily development'
                    })
        
        # Extract unresolved threads
        unresolved_match = re.search(r'\*\*Unresolved Threads:\*\*\s*\n(.*?)\n\n', content, re.DOTALL)
        if unresolved_match:
            unresolved_text = unresolved_match.group(1)
            for line in unresolved_text.split('\n'):
                if line.strip().startswith('-'):
                    thread = line.strip()[1:].strip()
                    data['unresolved_threads'].append({
                        'type': 'Future Work',
                        'item': thread,
                        'description': 'Carried forward from daily ledger'
                    })
        
        # Extract re-entry spell
        spell_match = re.search(r'### \*\*4\. Re‑entry Spell\*\*\s*\n> \*(.*?)\*', content, re.DOTALL)
        if spell_match:
            data['re_entry_spell'] = spell_match.group(1).strip()
        
        return data
    
    def generate_capsule_scroll(self, data):
        """Generate a capsule scroll from the provided data."""
        template = self.load_template()
        timestamp = datetime.now().isoformat()
        capsule_id = self.create_capsule_id(data['arc_name'], data.get('start_date'))
        
        # Replace template variables
        replacements = {
            '{ARC_NAME}': data['arc_name'],
            '{YYYY-MM-DD}': data.get('start_date', datetime.now().strftime('%Y-%m-%d')),
            '{SLUG}': re.sub(r'[^a-zA-Z0-9]+', '-', data['arc_name'].lower()).strip('-')[:30],
            '{TIMESTAMP}': timestamp,
            '{STATUS}': data.get('status', 'active'),
            '{START_DATE}': data.get('start_date', 'TBD'),
            '{END_DATE}': data.get('end_date', 'ongoing'),
            '{DURATION}': self.calculate_duration(data.get('start_date'), data.get('end_date')),
            '{THREE_SENTENCE_CONTEXT_RESET}': data.get('re_entry_spell', 'Context to be determined based on development arc progression. Key decisions and artifacts will guide future conversations. The adventure continues with preserved wisdom.'),
            '{TONE_KEYWORD_1}': 'collaborative',
            '{TONE_KEYWORD_2}': 'technical',
            '{TONE_KEYWORD_3}': 'iterative',
            '{TRIGGER_TYPE}': data.get('trigger_type', 'manual'),
            '{ISSUE_NUMBERS}': str(data.get('issue_numbers', [])),
            '{PR_NUMBERS}': str(data.get('pr_numbers', [])),
            '{TLDL_IDS}': str(data.get('tldl_ids', [])),
            '{UNRESOLVED_COUNT}': str(len(data.get('unresolved_threads', []))),
            '{VERSION}': '1.0.0'
        }
        
        content = template
        for placeholder, value in replacements.items():
            content = content.replace(placeholder, value)
        
        # Replace decision sections
        if data.get('core_decisions'):
            decisions_content = ""
            for i, decision in enumerate(data['core_decisions'][:3], 1):  # Limit to 3 decisions
                decisions_content += f"""### Decision {i}: {decision.get('title', 'Unnamed Decision')}
- **Context**: {decision.get('context', 'No context provided')}
- **Choice Made**: {decision.get('outcome', 'No outcome recorded')}
- **Rationale**: {decision.get('rationale', 'No rationale provided')}
- **Impact**: {decision.get('impact', 'Impact to be determined')}

"""
            # Replace the template decisions
            decision_pattern = r'### Decision 1:.*?(?=### Decision 2:)|### Decision 2:.*?(?=\*Add additional)'
            content = re.sub(decision_pattern, '', content, flags=re.DOTALL)
            content = content.replace('## 🎯 Core Decisions\n\n', f'## 🎯 Core Decisions\n\n{decisions_content}')
        
        # Replace artifacts sections
        if data.get('artifacts'):
            artifacts_content = "### Key References\n"
            for artifact in data['artifacts'][:5]:  # Limit to 5 artifacts
                artifacts_content += f"- {artifact}\n"
            artifacts_content += "\n"
            
            # Replace template artifacts
            artifact_pattern = r'### Issues.*?### TLDL Entries.*?\n\n'
            content = re.sub(artifact_pattern, artifacts_content, content, flags=re.DOTALL)
        
        # Replace glyphs section
        if data.get('glyphs'):
            glyphs_content = ""
            for glyph in data['glyphs'][:5]:  # Limit to 5 glyphs
                glyphs_content += f"- **{glyph.get('term', 'Unknown')}**: {glyph.get('definition', 'No definition provided')}\n"
            
            glyph_pattern = r'### Terminology.*?### Running Jokes.*?\n\n'
            content = re.sub(glyph_pattern, f"### Key Terms & References\n{glyphs_content}\n", content, flags=re.DOTALL)
        
        # Replace unresolved threads
        if data.get('unresolved_threads'):
            threads_content = ""
            for thread in data['unresolved_threads']:
                threads_content += f"- [ ] {thread.get('item', 'Unknown item')} - {thread.get('description', 'No description')}\n"
            
            threads_pattern = r'### Technical Debt.*?### Future Work.*?\n\n'
            content = content.replace('{{CORE_DECISIONS}}', decisions_content)
        
        # Replace artifacts section using placeholder
        if data.get('artifacts'):
            artifacts_content = "### Key References\n"
            for artifact in data['artifacts'][:5]:  # Limit to 5 artifacts
                artifacts_content += f"- {artifact}\n"
            artifacts_content += "\n"
            content = content.replace('{{ARTIFACTS}}', artifacts_content)
        
        # Replace glyphs section using placeholder
        if data.get('glyphs'):
            glyphs_content = ""
            for glyph in data['glyphs'][:5]:  # Limit to 5 glyphs
                glyphs_content += f"- **{glyph.get('term', 'Unknown')}**: {glyph.get('definition', 'No definition provided')}\n"
            content = content.replace('{{GLYPHS}}', f"### Key Terms & References\n{glyphs_content}\n")
        
        # Replace unresolved threads section using placeholder
        if data.get('unresolved_threads'):
            threads_content = ""
            for thread in data['unresolved_threads']:
                threads_content += f"- [ ] {thread.get('item', 'Unknown item')} - {thread.get('description', 'No description')}\n"
            content = content.replace('{{UNRESOLVED_THREADS}}', f"### Carried Forward\n{threads_content}\n")
        
        return content, capsule_id
    
    def calculate_duration(self, start_date, end_date):
        """Calculate duration between dates."""
        if not start_date or not end_date or end_date == 'ongoing':
            return 'ongoing'
        try:
            start = datetime.strptime(start_date, '%Y-%m-%d')
            end = datetime.strptime(end_date, '%Y-%m-%d')
            delta = end - start
            return f"{delta.days} days"
        except:
            return 'unknown'
    
    def save_capsule_scroll(self, content, capsule_id, status='active'):
        """Save the capsule scroll to the appropriate directory."""
        directory = self.active_dir if status == 'active' else self.archived_dir
        filename = f"{capsule_id}.md"
        filepath = os.path.join(directory, filename)
        
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        
        return filepath
    
    def create_from_daily_ledger(self, ledger_path):
        """Create a capsule scroll from a daily ledger."""
        data = self.extract_from_daily_ledger(ledger_path)
        content, capsule_id = self.generate_capsule_scroll(data)
        filepath = self.save_capsule_scroll(content, capsule_id)
        return filepath, capsule_id
    
    def create_manual(self, arc_name, start_date=None, end_date=None, status='active'):
        """Create a manual capsule scroll with basic information."""
        data = {
            'arc_name': arc_name,
            'start_date': start_date or datetime.now().strftime('%Y-%m-%d'),
            'end_date': end_date or 'ongoing',
            'status': status,
            'core_decisions': [],
            'artifacts': [],
            'glyphs': [],
            'unresolved_threads': [],
            're_entry_spell': f'Beginning of {arc_name} development arc. Initial context and decisions will be captured as the work progresses. The adventure starts here.'
        }
        
        content, capsule_id = self.generate_capsule_scroll(data)
        filepath = self.save_capsule_scroll(content, capsule_id, status)
        return filepath, capsule_id

def main():
    parser = argparse.ArgumentParser(description='Generate Capsule Scrolls for Archive Wall Continuity')
    parser.add_argument('--arc-name', help='Name of the development arc')
    parser.add_argument('--start-date', help='Start date (YYYY-MM-DD)')
    parser.add_argument('--end-date', help='End date (YYYY-MM-DD)')
    parser.add_argument('--status', choices=['active', 'archived'], default='active', help='Capsule status')
    parser.add_argument('--from-daily-ledger', help='Path to daily ledger to extract from')
    parser.add_argument('--project-root', help='Project root directory')
    
    args = parser.parse_args()
    
    if not args.arc_name and not args.from_daily_ledger:
        parser.error("Either --arc-name or --from-daily-ledger must be provided")
    
    generator = CapsuleScrollGenerator(args.project_root)
    
    try:
        if args.from_daily_ledger:
            filepath, capsule_id = generator.create_from_daily_ledger(args.from_daily_ledger)
            print(f"📜 Capsule Scroll created from daily ledger:")
        else:
            filepath, capsule_id = generator.create_manual(args.arc_name, args.start_date, args.end_date, args.status)
            print(f"📜 Manual Capsule Scroll created:")
        
        print(f"   ID: {capsule_id}")
        print(f"   Path: {filepath}")
        print(f"   Status: {args.status}")
        print()
        print("🧠 Next steps:")
        print("   1. Edit the capsule scroll to add specific context")
        print("   2. Link from TLDL entries for discoverability")
        print("   3. Reference in daily ledger for continuity")
        
    except Exception as e:
        print(f"❌ Error creating capsule scroll: {e}")
        sys.exit(1)

if __name__ == '__main__':
    main()